@here/harp-mapview/lib/FrustumIntersection.ts

```diff
@@ -0,0 +1,306 @@
+import { MapViewUtils, TileOffsetUtils } from "./Utils";
+
+/**
+ * Represents a unique TileKey and the area it takes up on screen.
+ *
+ * Note, in certain tiling projections, it is possible to have an offset, which represents a tile
+ * which has fully wrapped around, hence this defaults to 0 to simplify usage for projections which
+ * don't require it.
+ */
+export class TileKeyEntry {
+    constructor(public tileKey: TileKey, public area: number, public offset: number = 0) {}
+}
+
+function getGeoBox(tilingScheme: TilingScheme, childTileKey: TileKey, offset: number) {
+    const geoBox = tilingScheme.getGeoBox(childTileKey);
+    const longitudeOffset = 360.0 * offset;
+    geoBox.northEast.longitude += longitudeOffset;
+    geoBox.southWest.longitude += longitudeOffset;
+    return geoBox;
+}
+
+namespace FrustumIntersection {
+    export interface Result {
+        /**
+         * Tiles intersected by the frustum.
+         */
+        readonly tileKeyEntries: Map<number, TileKeyEntry>;
+        /**
+         * True if the intersection was calculated using precise elevation data, false if it's an
+         * approximation.
+         */
+        calculationFinal: boolean;
+    }
+}
+
+/**
+ * Computes the tiles intersected by the frustum defined by the current camera setup.
+ */
+export class FrustumIntersection {
+    private readonly m_frustum: THREE.Frustum = new THREE.Frustum();
+    // used to project global coordinates into camera local coordinates
+    private readonly m_viewProjectionMatrix = new THREE.Matrix4();
+    private readonly m_mapTileCuller: MapTileCuller;
+    private m_rootTileKeys: TileKeyEntry[] = [];
+    private readonly m_tileKeyEntries: Map<number, TileKeyEntry> = new Map();
+
+    constructor(
+        private readonly m_camera: THREE.PerspectiveCamera,
+        private readonly m_projection: Projection,
+        private readonly m_extendedFrustumCulling: boolean
+    ) {
+        this.m_mapTileCuller = new MapTileCuller(m_camera);
+    }
+
+    /**
+     * Updates the frustum to match the current camera setup.
+     */
+    updateFrustum() {
+        this.m_viewProjectionMatrix.multiplyMatrices(
+            this.m_camera.projectionMatrix,
+            this.m_camera.matrixWorldInverse
+        );
+
+        this.m_frustum.setFromMatrix(this.m_viewProjectionMatrix);
+
+        if (this.m_extendedFrustumCulling) {
+            this.m_mapTileCuller.setup();
+        }
+        this.computeRequiredInitialRootTileKeys(this.m_camera.position);
+    }
+
+    /**
+     * Computes the tiles intersected by the updated frustum, see [[updateFrustum]].
+     *
+     * @param tilingScheme The camera storage level, see [[MapView.storageLevel]].
+     * @param maxTileLevel The maximum tile level that will be checked for intersections.
+     * @param elevationRangeSource Source of elevation range data if any.
+     * @returns The computation result, see [[FrustumIntersection.Result]].
+     */
+    compute(
```

Comment:
Is there a need to have update and compute separate methods? In what use case would I call updateFrustum and not compute (and vice versa).

Reply:
updateFrustum needs to be done once per frame. computation is done as many times as different tile schemes.